### ISSUE TITLE
Tracking: Update `lightwalletd` RPCs for "spend before sync" algorithm

### ISSUE THREAD

**Issue by teor2345** [ZcashFoundation/zebra#6642]
## Motivation

The `zcashd` and `lightwalletd` teams have been working on a "spend before sync" algorithm, targeted for `zcashd` 5.6.0 in June 2023.

As part of that change, the JSON-RPC interface used by `lightwalletd` will need extra methods or fields.

Preparation:

<details>
<summary>completed preparation</summary>

- [x] #6746
- [x] #6955
- [x] #6859 (Update `incrementalmerkletree` crate to 0.4.0)
    - [x] #7143

- [x] #7292 
    - [x] #7390
    - [x] #7349

</details>

Zebra will need to implement some of those changes to remain compatible with the latest `lightwalletd` versions:
- [x] #6952
- [x] #6953
    - [x] #7532
    - [x] #7407
    - [x] #7570
- [x] #6954 
    - [x] #7449
        - [x] #7446
        - [x] #7450 
        - [x] #7451 
        - [x] #7513

Extra Dependencies & Tests:

<details>

<summary>completed</summary>

- [x] #7201
- [x] #6638
- [x] #7335

</details>

### Lower Priority

Minor bug or test fixes:
- [x] #7443
- [x] #7447 
- [x] #7461
- [x] #7444
- [x] #7351
- [x] #7463

<details>
<summary>completed performance & documentation</summary>

To get reasonable lookup performance:
- [x] #4784

Documentation:
- [x] #7215
- [x] #7319
    - [x] #7370

</details>

Optional Cleanups & Fixes:

<details>
<summary>these tickets are optional, ask Pili before starting them</summary>

Bug fixes:
- [ ] #7392
- [x] #7263 

Technical debt:
- [x] #7383
- [ ] #7377

Diagnostics:
- [x] #7369 
- [x] #7416

Needs decision/triage:
- [x] #7441

</details>

### Timing

`zcashd` aims to finalize the RPC design by the end of May, and then release it in the first two weeks of June.

`lightwalletd` will start calling the new RPCs around the same time, but they aim to be compatible with old RPC versions using an off-by-default flag.

### Implementation Notes

Zebra already stores the data we need for these lookups. We might need extra indexes for good performance, but that's not a breaking state change. (So we don't need to upgrade the state version, we can just automatically add/update a new column family with the new index at startup.)

We will need to implement #4784 to have acceptable note commitment tree scanning performance.

**Comment by mpguerra** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Engineering-Team-5fb24d9264a3e8000e666a9e/issues/ZcashFoundation/zebra/6642?planning-poker) with Zenhub @arya2 @conradoplg @dconnolly @oxarbitrage @teor2345 @upbqdn

**Comment by teor2345** [comment]
@mpguerra I can't estimate this ticket until the `zcashd` team has decided what the changes to the RPCs will be.

**Comment by str4d** [comment]
The current draft of the JSON-RPC changes is here (last few commits, the PR is currently based on another one and will be rebased as the backend changes are landed): https://github.com/zcash/zcash/pull/6677.

**Comment by teor2345** [comment]
I've added a draft list of RPC changes to the ticket, just trying to work out how to calculate subtree roots:
https://discord.com/channels/809218587167293450/809251012610752513/1110391483367960646

**Comment by teor2345** [comment]
Note from the engineering sync: This seems risky to do between the last release candidate and the first stable release. There is no specific deadline for this change.

**Comment by teor2345** [comment]
@mpguerra I'd like to talk through the design for this change with the engineers who will be working on it. We need to work out the tree API changes, and the database format changes.

Did you want to have a large group meeting? Or we could do it in two smaller groups: tree APIs and database format.

**Comment by teor2345** [comment]
Also I just split this task into tickets, can we schedule them in this sprint or next sprint?
I haven't set up the dependencies yet, but they are in the right order in the list in the PR.

**Comment by mpguerra** [comment]
Unscheduling as this will be the overall "tracking" issue and we'll schedule individual issues in sprints instead

**Comment by mpguerra** [comment]
> @mpguerra I'd like to talk through the design for this change with the engineers who will be working on it. We need to work out the tree API changes, and the database format changes.
> 
> Did you want to have a large group meeting? Or we could do it in two smaller groups: tree APIs and database format.

I think it's probably best to have smaller group meetings and involve only the people that will need to be involved

**Comment by mpguerra** [comment]
We are done here for now 🎉
